common/src/main/java/io/netty/util/internal/PlatformDependent.java

```diff
@@ -496,6 +499,51 @@ public static void copyMemory(long srcAddr, byte[] dst, int dstIndex, long lengt
         PlatformDependent0.copyMemory(null, srcAddr, dst, BYTE_ARRAY_BASE_OFFSET + dstIndex, length);
     }
 
+    public static ByteBuffer allocateDirectNoCleaner(int capacity) {
+        if (DIRECT_MEMORY_COUNTER != null) {
+            for (;;) {
+                long usedMemory = DIRECT_MEMORY_COUNTER.get();
+                long newUsedMemory = usedMemory + capacity;
+                if (newUsedMemory > maxDirectMemory()) {
+                    throw new OutOfMemoryError("No more memory available");
```

Comment:
`"failed to allocate " + capacity + " byte(s) of direct memory (used: " + usedMemory + ", max: " + maxDirectMemory() + ')'`